### PR TITLE
Add inventory recommendation on how to loop over hosts

### LINKED ISSUE #26: Add recommendation to not loop over hosts but use the inventory
### Anti-pattern:

```yaml
list_of_hosts_to_create:
  host1: var1, var2, var3
  host2: var1, var2, var3
```

then loop over the list and (often) create the host against oVirt, vSphere, etc...

### Recommended pattern:

```
inventory/host_vars/{host1,host2}/vm_create.yml (with var1, var2, var3 + variable vm_creation_server)
```

then:

```yaml
- hosts: vms_to_be_created
  gather_facts: false  # important or it will fail if the VM doesn't exist yet
  tasks:
    - name: create VM(s)
       create_vm:
          vm_creation_endpoint: "{{ vm_creation_server }}"
          [ ... etc ... ]
```

Rationale: is parallelized automatically by Ansible, easier to maintain, easier to `--limit` to certain hosts

### PR DESCRIPTION
**Pull request by ericzolf**
Closes #26

### CODE REVIEW

File: inventories/README.adoc [hunk 1]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
             └── locations.yml
 ----
 ====
+
+== Rely on your inventory to loop over hosts, don't create lists of hosts
+[%collapsible]
+====
+Explanations::
+When there is the need to do the same task on multiple hosts, don't create a variable with a list of hosts and loop over it.
```
Comment (mophahr) [inline inventories/README.adoc:165]:
```suggestion
To perform the same task on multiple hosts, don't create a variable with a list of hosts and loop over it.
```

File: inventories/README.adoc [hunk 2]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
+Instead use as much as possible the capabilities of your inventory, which is already a kind of list of hosts.
++
+The anti-pattern is especially seen when it comes to the provisioning of hosts against some kind of manager (e.g. a list of VMs against an hypervisor manager, like oVirt/RHV or vCenter, or against a management tool like Foreman/Satellite, or our beloved AWX/Tower/controller).
```
Comment (mophahr) [inline inventories/README.adoc:168]:
```suggestion
The anti-pattern is especially obvious in the example of provisioning hosts on some kind of manager.
Commonly seen automation tasks of this kind are spinning up a list of VMs via a hypervisor manager like oVirt/RHV or vCenter, or calling a management tool like Foreman/Satellite or even our beloved AWX/Tower/controller.
```

File: inventories/README.adoc [hunk 3]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
+
+Rationale::
+There are 4 main reasons for following this advice:
++
+. a list of hosts is more difficult to maintain than an inventory structure, and tends to become very quickly difficult to oversee.
```
Comment (mophahr) [inline inventories/README.adoc:173]:
```suggestion
. a list of hosts is more difficult to maintain than an inventory structure, and it tends to become very difficult to oversee very quickly .
```

File: inventories/README.adoc [hunk 4]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
+It is especially true as you generally need to maintain your hosts in your inventory.
```
Comment (mophahr) [inline inventories/README.adoc:174]:
```suggestion
This is especially true as you generally already need to maintain your hosts in your inventory.
```

File: inventories/README.adoc [hunk 5]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
+This brings us to the 2nd advantage:
+. you avoid duplicating information, as you often need the same kind of information in your inventory that you also need in order to provision your VMs.
+In your inventory, you can also use groups to define group variables, automatically inherited by hosts.
+You can try to implement a similar inheritance pattern with your list of hosts, but it becomes difficult and hand-crafted.
```
Comment (mophahr) [inline inventories/README.adoc:178]:
```suggestion
You can try to implement a similar inheritance pattern with your list of hosts, but it quickly becomes difficult and _hand-crafted_.
```

File: inventories/README.adoc [hunk 6]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
+. as you loop through the hosts of an inventory, Ansible helps you with https://docs.ansible.com/ansible/latest/user_guide/playbooks_strategies.html[parallelization, throttling, etc], all kind of things you can't do with your own list (well, you can combine https://docs.ansible.com/ansible/latest/user_guide/playbooks_async.html[async and loop] to reach something like this, but it's a lot more complex to handle).
```
Comment (mophahr) [inline inventories/README.adoc:179]:
```suggestion
. as you loop through the hosts of an inventory, Ansible helps you with https://docs.ansible.com/ansible/latest/user_guide/playbooks_strategies.html[parallelization, throttling, etc], all of which you can't do easily with your own list (technically, you _can_ combine https://docs.ansible.com/ansible/latest/user_guide/playbooks_async.html[async and loop] to reach something like this, but it's a lot more complex to handle than letting Ansible do the heavy lifting for you).
```

File: inventories/README.adoc [hunk 7]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
++
+.The "good" way to loop over hosts
+[source,yaml]
+----
+include::inventory_loop_hosts/playbook_good.yml[]
+----
++
+The result isn't overwhelming in this simple setup but you would of course better appreciate if the provisioning would take half an hour instead of a fraction of seconds:
++
+.Comparaison of the execution times between good and bad
```
Comment (mophahr) [inline inventories/README.adoc:239]:
```suggestion
.Comparison of the execution times between the "good" and the "bad" implementation
```

File: inventories/README.adoc [hunk 8]
```diff
@@ -157,3 +157,100 @@ inventory_satellite/
...
+create some file to simulate an API call to provision a host ------------ 0.86s
+$ ANSIBLE_STDOUT_CALLBACK=profile_tasks \
+	ansible-playbook -i inventory_good playbook_good.yml 
+Saturday 23 October 2021  13:11:55 +0200 (0:00:00.040)       0:00:00.040 ****** 
+Saturday 23 October 2021  13:11:56 +0200 (0:00:00.569)       0:00:00.610 ****** 
+=============================================================================== 
+create some file to simulate an API call to provision a host ------------ 0.57s
+----
++
+TIP: if for some reason, you can't follow the recommendation, you can at least avoid duplicating too much information by indirectly referencing the hosts' variables as in `"{{ hostvars[item.name]['provision_value'] }}"`. Not so bad...
```
Comment (mophahr) [inline inventories/README.adoc:255]:
```suggestion
TIP: if for some reason, you can't follow the recommendation, you can at least avoid duplicating too much information by indirectly referencing the hosts' variables as in `"{{ hostvars[item.name]['provision_value'] }}"`.
```

**Review by mophahr — COMMENTED** [review]
Only cosmetic suggestions and a few typos. Nice write-up!

File: inventories/README.adoc (hunk 3 above)
Comment (ericzolf) [reply inventories/README.adoc:173]:
```suggestion
. a list of hosts is more difficult to maintain than an inventory structure, and tends it quickly to become difficult to oversee.
```

File: inventories/README.adoc (hunk 7 above)
Comment (ericzolf) [reply inventories/README.adoc:239]: Aah, my French shining through...

File: inventories/README.adoc (hunk 8 above)
Comment (ericzolf) [reply inventories/README.adoc:255]: That was supposed to be a funny reference to the 3rd code example. Not so funny?

**Review by mophahr — APPROVED** [review]
👍

**Comment by ericzolf** [comment]
Approved during 2021-11-17 CoP meeting

**Pull request merged**